source/v5/mqtt3_to_mqtt5_adapter.c

```diff
@@ -1651,7 +1681,549 @@ static uint16_t s_aws_mqtt_client_connection_5_publish(
+
+        for (size_t i = 0; i < suback->reason_code_count; ++i) {
+            struct aws_mqtt_topic_subscription *subscription = subscription_ptr + i;
+            AWS_ZERO_STRUCT(*subscription);
+
+            subscription->qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[i]);
+
+            if (i < subscription_count) {
+                struct aws_mqtt_subscription_set_subscription_record *record = NULL;
+                aws_array_list_get_at(&subscribe_op->subscriptions, &record, i);
+
+                subscription->topic = record->subscription_view.topic_filter;
+                subscription->on_publish = record->subscription_view.on_publish_received;
+                subscription->on_publish_ud = record->subscription_view.callback_user_data;
+                subscription->on_cleanup = record->subscription_view.on_cleanup;
+            }
+
+            aws_array_list_push_back(&multi_sub_list, &subscription);
+        }
+
+        (*subscribe_op->on_multi_suback)(
+            &adapter->base,
+            subscribe_op->base.id,
+            &multi_sub_list,
+            error_code,
+            subscribe_op->on_multi_suback_user_data);
+    }
+
+    aws_mqtt3_to_mqtt5_adapter_operation_table_remove_operation(
+        &subscribe_op->base.adapter->operational_state, subscribe_op->base.id);
+}
+
+static int s_validate_adapter_subscribe_options(const struct aws_mqtt3_to_mqtt5_adapter_subscribe_options *options) {
+    if (options->subscription_count == 0) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    for (size_t i = 0; i < options->subscription_count; ++i) {
+        struct aws_mqtt_topic_subscription *subscription_options = &options->subscriptions[i];
+        enum aws_mqtt_qos qos = subscription_options->qos;
+        if (qos < AWS_MQTT_QOS_AT_MOST_ONCE || qos > AWS_MQTT_QOS_EXACTLY_ONCE) {
+            return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+struct aws_mqtt3_to_mqtt5_adapter_operation_subscribe *aws_mqtt3_to_mqtt5_adapter_operation_new_subscribe(
+    struct aws_allocator *allocator,
+    const struct aws_mqtt3_to_mqtt5_adapter_subscribe_options *options) {
+
+    if (s_validate_adapter_subscribe_options(options)) {
+        return NULL;
+    }
+
+    struct aws_mqtt3_to_mqtt5_adapter_operation_subscribe *subscribe_op =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_mqtt3_to_mqtt5_adapter_operation_subscribe));
+
+    subscribe_op->base.allocator = allocator;
+    aws_ref_count_init(&subscribe_op->base.ref_count, subscribe_op, s_adapter_subscribe_operation_destroy);
+    subscribe_op->base.impl = subscribe_op;
+    subscribe_op->base.type = AWS_MQTT3TO5_AOT_SUBSCRIBE;
+    subscribe_op->base.adapter = options->adapter;
+    subscribe_op->base.holding_adapter_ref = false;
+
+    /* make persistent adapter sub array */
+    aws_array_list_init_dynamic(
```

Comment:
`aws_array_list_init_dynamic` looks like it can fail? Or at least that possibility is handled in a lot of other areas. Should we check its result and goto error here?

Reply:
Memory allocation only